daemon/api.go

```diff
@@ -585,21 +585,50 @@ func deleteOp(c *Command, r *http.Request) Response {
 	}
 }
 
+type licT struct {
+	Intro   string
+	License string
+	Agreed  bool
+}
+
+func (*licT) Error() string {
+	// licT isn't an error, strictly speaking, but can
+	// be converted into one for convenience
+	return "license agreement required"
+}
+
 type packageInstruction struct {
+	progress.NullProgress
 	Action   string `json:"action"`
 	LeaveOld bool   `json:"leave_old"`
+	License  *licT  `json:"license"`
 	pkg      string
-	prog     progress.Meter
 }
 
+func (inst *packageInstruction) Agreed(intro, license string) bool {
+	if inst.License == nil || !inst.License.Agreed || inst.License.Intro != intro || inst.License.License != license {
+		inst.License = &licT{Intro: intro, License: license, Agreed: false}
+		return false
+	}
+
+	return true
+}
+
+var snappyInstall = snappy.Install
+
 func (inst *packageInstruction) install() interface{} {
 	flags := snappy.DoInstallGC
 	if inst.LeaveOld {
 		flags = 0
 	}
-	_, err := snappy.Install(inst.pkg, flags, inst.prog)
-
+	_, err := snappyInstall(inst.pkg, flags, inst)
 	if err != nil {
+		// to be more rigiorous we could also check
+		// err, ok := err.(*snappy.ErrInstallFailed); ok && err.OrigErr == snappy.ErrLicenseNotAccepted
+		// but that seems a bit overkill
```

Comment:
It seems overkill because there's no easy way to check the error. If it was something as cheap as `snappy.IsLicensePending(err)`, which feels reasonable, it might be more tempting to avoid the heuristics which turn _every_ error returned by snappy.Install into a license error based on an assumption built on knowledge of the current version of that function.